forth_src/io.fs

```diff
@@ -0,0 +1,79 @@
+require open
+
+\ Set input device to open file #
+\ returns 0 on success, file # on error
+code chkin? ( file# -- file# result )
+w stx,
+lsb lda,x tax, \ x = file#
+$ffc6 jsr, \ CHKIN
++branch bcs, \ carry set = error
+0 lda,# \ A is only valid on error
+:+
+w ldx, dex, 
+lsb sta,x
+0 lda,# msb sta,x
+;code
+
+\ Set output device to open file #
+\ returns 0 on success, file # on error
+code chkout? ( file# -- file# result )
+w stx,
+lsb lda,x tax, \ x = file#
+$ffc9 jsr, \ CHKOUT
++branch bcs, \ carry set = error
+0 lda,# \ A is only valid on error
+:+
+w ldx, dex,
+lsb sta,x
+0 lda,# msb sta,x
+;code
+
+\ Reset input and output to console 
+code clrchn ( -- ) 
+txa, pha,
+$ffcc jsr,  \ CLRCH
+pla, tax,
+;code
+
+\ Read status of last IO operation
+code readst ( -- status )
+dex, 0 lda,# msb sta,x
+$ffb7 jsr, \ READST
+lsb sta,x
+;code
+
+\ Get a byte from input device
+code chrin ( -- chr )
+dex, w stx, 0 lda,# msb sta,x
+$ffcf jsr, \ CHRIN
+w ldx, lsb sta,x
+;code
+
+\ handle return value from open?, chkin?
+\ and chkout?. If result is nonzero,
+\ close file, print error msg, and abort
+: ioerr ( file# result -- )
+?dup if rvs case
+2 of ." file# in use" endof
+3 of ." file not open" endof
+5 of ." device not present" endof
+6 of ." not input file" endof
+7 of ." not output file" endof
+." io err" 
+endcase clrchn close cr abort
+else drop then ;
+
+\ Open a file
+\ 'easy' version, aborts on error
+: open ( caddr u file# sa -- )
+open? ioerr ;
+
+\ Set input device to open file #
+\ 'easy' version, aborts on error
+: chkin ( file# -- ) 
+chkin? ioerr ;
+
+\ Set output device to open file #
+\ 'easy' version, aborts on error
+: chkout ( file# -- )
+chkout? ioerr ;
```

Comment:
I think everything looks kind of good, the one thing I'm not sure about are these "easy" versions. Maybe it would make things in practice simpler, to only have one set of open/chkin/chkout words that always return a status?
When looking in C and Forth libraries, this to me seems like the normal thing to do.

I think cc65 is a good reference: https://cc65.github.io/doc/funcref.html#ss2.11

For those who would like an "easy" version without error handling, it could always be added later?
What do you think?

Reply:
As a contrary reference, 64Forth OPEN/CHKIN/CHKOUT seem to have in-built error handling - or at least those functions do not return any value to indicate success or failure.

Forth is kind of difficult because the only thing you know, in ease of use and performance it is "supposed" to be somewhere between BASIC and C - and that leaves quite a lot of space for tradeoffs and design choices.

Reply:
I think what definitely speaks for having `ioerr` around is that the printer example now looks simple and good.

Changing printer example to e.g. `0 0 7 47 open ioerr ( open address 7 on device 4 as file 47 )` would make it a bit less nice, but it's still OK.

Leaving all error handling to be implemented by the user, then it feels like the printer example will not be nice anymore.

I would consider renaming `ioerr` to `ioabort`. I think it becomes a bit easier to understand what it does in that way.

Reply:
I am sorry for all the "loud thinking" but my current opinion is, as a summary:

 - drop the "easy" versions, only keep one set of `open`/`chkin`/`chout`
 - rename `ioerr` to `ioabort` and document it
 - in documentation, rename "result" of above word signatures to "ioresult" to clarify that these words are supposed to be used together

Reply:
I like this idea a lot. `open ioabort` isn't *really* that much more to type than just `open`. I think it hits a nice middle ground of 'more expressive than BASIC, but easier than asm or C'.